doc/neps/nep-0033-array-creation-dispatch-with-array-function.rst

```diff
@@ -0,0 +1,184 @@
+===========================================================
+
+:Author: Peter Andreas Entschev <pentschev@nvidia.com>
+:Status: Draft
+:Type: Standards Track
+:Created: 2019-10-15
+:Updated: 2019-10-15
+:Resolution:
+
+Abstract
+--------
+
+We propose the introduction of a new keyword argument ``like=`` to all array
+creation functions to permit dispatching of such functions by the
+``__array_function__`` protocol, addressing one of the protocol shortcomings,
+as described by NEP-18 [1]_.
+
+Detailed description
+--------------------
+
+The introduction of the ``__array_function__`` protocol allowed downstream
+library developers to use NumPy as a dispatching API. However, the protocol
+did not -- and did not intend to -- address the creation of arrays by downstream
+libraries, preventing those libraries from using such important functionality in
+that context.
+
+Other NEPs have been written to address parts of that limitation, such as the
+introduction of the ``__duckarray__`` protocol in NEP-30 [2]_, and the
+introduction of an overriding mechanism called ``uarray`` by NEP-31 [3]_.
+
+The purpose of this NEP is to address that shortcoming in a simple and
+straighforward way: introduce a new ``like=`` keyword argument, similar to how
+the ``empty_like`` family of functions work. When array creation functions
+receive such an argument, they will trigger the ``__array_function__`` protocol,
+and call the downstream library's own array creation function implementation.
+The ``like=`` argument, as its own name suggests, shall be used solely for the
+purpose of identifying where to dispatch.  In contrast to the way
+``__array_function__`` has been used so far (the first argument identifies where
+to dispatch), and to avoid breaking NumPy's API with regards to array creation,
+the new ``like=`` keyword shall be used for the purpose of dispatching.
+
+Usage Guidance
+~~~~~~~~~~~~~~
+
+The new ``like=`` keyword is solely intended to identify the downstream library
+where to dispatch and the object is used only as reference, meaning that no
+modifications, copies or processing will be performed on that object.
+
+We expect that this functionality will be mostly useful to library developers,
+allowing them to create new arrays for internal usage based on arrays passed
+by the user, preventing unnecessary creation of NumPy arrays that will
+ultimately lead to an additional conversion into a downstream array type.
+
+Implementation
+--------------
+
+The implementation requires introducing a new ``like=`` keyword to all existing
+array creation functions of NumPy. As of the writing of this NEP, a complete
+list of array creation functions can be found in [4]_.
+
+This newly proposed keyword shall be removed by the ``__array_function__``
+mechanism from the keyword dictionary before dispatching. The purpose for this
+is twofold:
+
+1. The object will have no use in the downstream library's implementation; and
```

Comment:
This part feels questionable to me. I don't think it's a good idea to add special cases to a protocol, so I think it's a good idea to include `like` in argument even if it wouldn't be useful.

I can also definitely imagine a library with multiple duck array types (e.g., pydata/sparse) that uses a single implementation that relies upon the `like` argument.

Reply:
In that case, we could just have a separate `__array_function__` on each of our types. But in any case, I would rather the `like=` kwarg was removed, it'd drive adaptability up quickly... But _only_ for array creation functions, not others.

Reply:
> I can also definitely imagine a library with multiple duck array types (e.g., pydata/sparse) that uses a single implementation that relies upon the like argument.

But that would still rely on the NumPy dispatching system, no? In that case, I would expect the implementation to find out internally what is the actual type it needs to create and call NumPy's API to create a new object, much like what we do now with `_meta` in Dask.

Reply:
> In that case, we could just have a separate __array_function__ on each of our types.

I'm not sure I understand your comment, but perhaps this is what I also tried to say in my comment just above?

> But in any case, I would rather the like= kwarg was removed, it'd drive adaptability up quickly... But only for array creation functions, not others.

This is indeed my intent, I think the removal would be incredibly helpful for downstream libraries, even though I see why @shoyer is concerned about it. Perhaps would be nice to have more opinions here and see if they eventually converge in some direction.

Reply:
> I'm not sure I understand your comment, but perhaps this is what I also tried to say in my comment just above?

I mean that PyData/Sparse has different types for the different kinds of arrays it uses. Each of those types could have its own `__array_function__` if we wanted to specialize array creation functions at all, instead of using one overarching `__array_function__`.

Reply:
However I could think of `da.Array` using such a keyword to copy e.g. chunk sizes and inner array types from the source array.

Reply:
> I mean that PyData/Sparse has different types for the different kinds of arrays it uses. Each of those types could have its own __array_function__ if we wanted to specialize array creation functions at all, instead of using one overarching __array_function__.

Oh, I see now. Yeah, that would be an alternative, but that could be a big burden too. So maybe in that case just not removing `like=` kwarg would make more sense?

> However I could think of da.Array using such a keyword to copy e.g. chunk sizes and inner array types from the source array.

In the case of Dask, we can now use `_meta` to identify the inner array and use it as the `like` object. So there I think this wouldn't be a problem.

Reply:
True, you would get the `like=` array in the form of `self`.

Reply:
> but that could be a big burden too

Not really, I consider it a separation of concerns.

Reply:
You're right, this makes sense. I was mostly saying this from the quick/seamless adaptability point-of-view.